File: api/src/main/java/net/blueshell/api/validation/survey/ValidQuestionListValidator.java
package net.blueshell.api.validation.survey;

import jakarta.validation.ConstraintValidator;
import jakarta.validation.ConstraintValidatorContext;
import net.blueshell.api.dto.survey.QuestionDTO;

import java.util.HashSet;
import java.util.List;
import java.util.Set;

public class ValidQuestionListValidator implements ConstraintValidator<ValidQuestionList, List<QuestionDTO>> {

    @Override
    public boolean isValid(List<QuestionDTO> questions, ConstraintValidatorContext context) {
        if (questions == null || questions.isEmpty()) {
            // handled by @NotEmpty
            return true;
        }

        Set<Long> seenIdx = new HashSet<>();
        for (QuestionDTO q : questions) {
            if (q.getIdx() == null) return false;
            if (!seenIdx.add(q.getIdx())) {
                context.disableDefaultConstraintViolation();
                context.buildConstraintViolationWithTemplate("Duplicate question index: " + q.getIdx())
                        .addConstraintViolation();
                return false;
            }
        }

        return true;
    }
}

File: api/src/main/kotlin/net/blueshell/api/validation/survey/ValidQuestionListValidator.kt
package net.blueshell.api.validation.survey

import jakarta.validation.ConstraintValidator
import jakarta.validation.ConstraintValidatorContext
import net.blueshell.api.dto.survey.QuestionDTO

class ValidQuestionListValidator :
    ConstraintValidator<ValidQuestionList, MutableList<QuestionDTO>> {
    override fun isValid(questions: MutableList<QuestionDTO>?, context: ConstraintValidatorContext): Boolean {
        if (questions.isNullOrEmpty()) {
            // handled by @NotEmpty
            return true
        }

        val seenIdx: MutableSet<Long?> = HashSet()
        for (q in questions) {
            if (q.idx == null) return false
            if (!seenIdx.add(q.idx)) {
                context.disableDefaultConstraintViolation()
                context.buildConstraintViolationWithTemplate("Duplicate question index: " + q.idx)
                    .addConstraintViolation()
                return false
            }
        }

        return true
    }
}

